src/nbl/ext/FFT/FFT.cpp

```diff
@@ -0,0 +1,368 @@
+
+layout(local_size_x=_NBL_GLSL_EXT_FFT_BLOCK_SIZE_X_DEFINED_, local_size_y=_NBL_GLSL_EXT_FFT_BLOCK_SIZE_Y_DEFINED_, local_size_z=_NBL_GLSL_EXT_FFT_BLOCK_SIZE_Z_DEFINED_) in;
+ 
+#define _NBL_GLSL_EXT_FFT_MAX_DIM_SIZE_ %u
+#define _NBL_GLSL_EXT_FFT_MAX_ITEMS_PER_THREAD %u
+ 
+#define _NBL_GLSL_EXT_FFT_GET_DATA_DEFINED_
+#define _NBL_GLSL_EXT_FFT_SET_DATA_DEFINED_
+#include "nbl/builtin/glsl/ext/FFT/fft.glsl"
+
+// Input Descriptor
+
+struct nbl_glsl_ext_FFT_input_t
+{
+	vec2 complex_value;
+};
+
+#ifndef _NBL_GLSL_EXT_FFT_INPUT_SET_DEFINED_
+#define _NBL_GLSL_EXT_FFT_INPUT_SET_DEFINED_ 0
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_INPUT_BINDING_DEFINED_
+#define _NBL_GLSL_EXT_FFT_INPUT_BINDING_DEFINED_ 0
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_INPUT_DESCRIPTOR_DEFINED_
+
+#if USE_SSBO_FOR_INPUT > 0
+#define _NBL_GLSL_EXT_FFT_INPUT_DESCRIPTOR_DEFINED_
+layout(set=_NBL_GLSL_EXT_FFT_INPUT_SET_DEFINED_, binding=_NBL_GLSL_EXT_FFT_INPUT_BINDING_DEFINED_) readonly restrict buffer InputBuffer
+{
+	nbl_glsl_ext_FFT_input_t inData[];
+};
+#else 
+layout(set=_NBL_GLSL_EXT_FFT_INPUT_SET_DEFINED_, binding=_NBL_GLSL_EXT_FFT_INPUT_BINDING_DEFINED_) uniform sampler2D inputImage;
+#endif
+
+#endif
+
+// Output Descriptor
+
+struct nbl_glsl_ext_FFT_output_t
+{
+	vec2 complex_value;
+};
+
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_ 0
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_ 1
+#endif
+
+#ifndef _NBL_GLSL_EXT_FFT_OUTPUT_DESCRIPTOR_DEFINED_
+#define _NBL_GLSL_EXT_FFT_OUTPUT_DESCRIPTOR_DEFINED_
+layout(set=_NBL_GLSL_EXT_FFT_OUTPUT_SET_DEFINED_, binding=_NBL_GLSL_EXT_FFT_OUTPUT_BINDING_DEFINED_) restrict buffer OutputBuffer
+{
+	nbl_glsl_ext_FFT_output_t outData[];
+};
+#endif
+
+// Get/Set Data Function
+
+vec2 nbl_glsl_ext_FFT_getData(in uvec3 coordinate, in uint channel)
+{
+	vec2 retValue = vec2(0, 0);
+#if USE_SSBO_FOR_INPUT > 0
+	uvec3 dimension = pc.dimension;
+	uint index = channel * (dimension.x * dimension.y * dimension.z) + coordinate.z * (dimension.x * dimension.y) + coordinate.y * (dimension.x) + coordinate.x;
+	retValue = inData[index].complex_value;
+#else
+	vec4 texelValue= texelFetch(inputImage, ivec2(coordinate.xy), 0);
+	retValue = vec2(texelValue[channel], 0.0f);
+#endif
+	return retValue;
+}
+
+void nbl_glsl_ext_FFT_setData(in uvec3 coordinate, in uint channel, in vec2 complex_value)
+{
+	uvec3 dimension = pc.padded_dimension;
+	uint index = channel * (dimension.x * dimension.y * dimension.z) + coordinate.z * (dimension.x * dimension.y) + coordinate.y * (dimension.x) + coordinate.x;
+	outData[index].complex_value = complex_value;
+}
+
+void main()
+{
+	nbl_glsl_ext_FFT();
+}
+
+)===";
+
+	const size_t extraSize = 32 + 32 + 32 + 32;
+
+	const uint32_t maxItemsPerThread = core::ceil(float(maxPaddedDimensionSize) / DEFAULT_WORK_GROUP_X_DIM);
```

Comment:
dont use floats to achieve roundup.... also if we optimized the smem usage then we'd probably not need to give the max number of items per thread at all